### ISSUE TITLE
Doubts on final loss computation while running DistributedDataParallel

### ISSUE THREAD

**Issue by jerinphilip** [pytorch/fairseq#779]
I'm trying to understand how distributed training works in this library. There are two places where `loss.backward()` and `optimizer.step()` seems to be called. The first is in `task.train_step(...)`, another in `trainer.train_step(..)` currently which are linked below. Which of the two is being used?

The loss computed in `task.train_step(...)` I think corresponds to the minibatch from the worker's shard. Where is the `all-gather` and dividing by total batch size happening? Is this not required?

The `trainer.train_step(...)` multiplies grads by `distributed_world_size/sample_size`. Can someone tell me what's happening under the hood?

**Task Backprop**
https://github.com/pytorch/fairseq/blob/5408bc0821144235e15f59ed7de1ca8db0e297f9/fairseq/tasks/fairseq_task.py#L210-L235

**Trainer backprop**
https://github.com/pytorch/fairseq/blob/5408bc0821144235e15f59ed7de1ca8db0e297f9/fairseq/trainer.py#L255-L258

https://github.com/pytorch/fairseq/blob/5408bc0821144235e15f59ed7de1ca8db0e297f9/fairseq/trainer.py#L322-L329

**Comment by myleott** [comment]
Your understanding is correct. We use torch’s DistributedDataParallel to handle communication of gradients across workers. This happens transparently in the backward pass.

Thereafter, Trainer.train_step is the only place where optimizer.step is called, and it handles renormalizing the gradients based on the sample_size defined by the criterion. The multiplication by distributed_world_size/sample_size is to correct the built-in averaging that happens in torch DistributedDataParallel.
